### DISCUSSION TITLE
Add `xmlwriter` extension

### DISCUSSION THREAD

**Discussion by DanielVera987** [NativePHP/laravel#309] [Ideas]
### What were you trying to do?

I have a system so that I can read xml file and export them as excel, the problem is that at the time of reading xml files appears me that is missing extension, how can I urge it?

### What happened?

After reading the N xml is exported in an excel that I make use of the larave-excel library and can download the file

### How to reproduce the bug

![Captura de pantalla 2024-05-15 a la(s) 10 48 24 a m](https://github.com/NativePHP/laravel/assets/59943431/401c144c-659d-441d-80f7-0cbe03297c6b)

### Package Versions

0.6.3

### PHP Version

8.2.0

### Laravel Version

10.10

### Node Version

16.10.0

### OS version

macos 14.4.1

**Comment by simonhamp** [comment]
Hey @DanielVera987. The `xmlwriter` extension isn't compiled into the binaries currently.

It looks like there's wide support for it across all platforms, so it would just be a case of compiling it in.

But I'm not sure if this makes sense for the central binaries we ship though, so I'm moving this to the Ideas board so we can gather feedback on how much need there is for it.

**Reply by ohnotnow** [reply]
Hi,

Came across this while trying to also create an app the processes excel files.  But it needs the xml writer extension and zip support too.  I can see the '-bulk' version of static php has it but not the one that is installed in my nativephp app.  I'd guess mangling excel will be a relatively common thing - so if there's a chance of support (or a way for me to pick a different php build?) I'd be delighted.  The experience so far has been great!

Edit: just to add for reference, these are the extensions the phpspreadsheet package needs :

php_zip
php_xml
php_gd2
php_iconv
php_simplexml
php_xmlreader
php_zlib

**Reply by poldixd** [reply]
I have encountered the same problem. I want to generate an Excel file with maatwebsite/excel and this does not work without xmlwriter. Almost all excel packages need xmlwriter.

**Reply by simonhamp** [reply]
Yep. I recommend you build your own PHP binaries with the necessary extensions

**Reply by chack1172** [reply]
@simonhamp I tried building my own PHP binaries adding xmlwriter, but running native:serve does not start the app.
Looking at laravel.log I'm getting: Invalid crypto_method request option: TLS 1.2 not supported by your version of cURL.
It is not related to xmlwriter, did I do something wrong?
I forked nativephp/php-bin, added xmlwriter to php-extensions.txt and manually run the Build PHP action.
PHP 8.4, Windows x64

**Reply by DanielVera987** [reply]
@simonhamp some news on the subject?

**Reply by simonhamp** [reply]
@DanielVera987 you will still need to [compile your own binaries](https://nativephp.com/docs/desktop/1/digging-deeper/php-binaries).

**Reply by PeteBishwhip** [reply]
> @simonhamp I tried building my own PHP binaries adding xmlwriter, but running native:serve does not start the app. Looking at laravel.log I'm getting: Invalid crypto_method request option: TLS 1.2 not supported by your version of cURL. It is not related to xmlwriter, did I do something wrong? I forked nativephp/php-bin, added xmlwriter to php-extensions.txt and manually run the Build PHP action. PHP 8.4, Windows x64

We're unable to provide support with custom binary builds however, it sounds like your binary is using an older version of Curl. You'll need to rectify that.

**Reply by kyagie** [reply]
Setting the path for custom binaries seems not to work as well. What other ways can this be done.

**Comment by kyagie** [comment]
@DanielVera987 @PeteBishwhip Please add the xmlwriter extension. We are having little success with compiling it in the custom binary builds

**Comment by uluumbch** [comment]
any update on this? or maybe any guide to build the custom with php xmlwriter enabled?

**Comment by Sethisme99** [comment]
is there any update on this problem yet?

**Comment by Lea-Reift** [comment]
# EDIT
This happened when I retried the compilation with my steps. It didn't work, but not because of the steps, but because the script didn't decompress the libraries after downloading. If this happens, you have to unzip them manually and then try the build again.

@uluumbch @Sethisme99 and everyone who needs a more detailed instruction on the build process, here is what I did:
## clone repo
```bash
git clone https://github.com/crazywhalecc/static-php-cli.git --depth=1
cd static-php-cli
```

# IMPORTANT
ALL THESE COMMANDS WERE RUN IN THE `static-php-cli` DIR

## Setup runtime
This will generate a php binary (with the minimun version required for the build) and the composer binary (I recommend using these instead)
```bash
bin/setup-runtime
```

## You need to install the PHP environment first before running Composer and this project. The installation method can be referred to below.
```bash
./runtime/php.exe ./runtime/composer.phar update
```

## Prepare dependencies
The docs of static-php-cli set the order as:
1. download
1. doctor

But what worked for me was running them like this:
```bash
./runtime/php.exe ./bin/spc doctor --auto-fix

#then
./runtime/php.exe ./bin/spc download php-src \ #note the `php-src` argument
    --with-php={desired_php_version} \
    --for-extensions={list_of_extensions_comma_separated}
    --for-libs={list_of_libs_comma_separated} //or directly --all and you save the extensions and libs options
```

## Build command

Now you can build your php binary as desired, with the format in the docs.
```bash
./runtime/php.exe ./bin/spc build \
    "{list_of_extensions_comma_separated}" \
    --build-cli \ # or the target you need
    --with-libs="{list_of_libs_comma_separated}"
    -I "{php_ini_variable_assignation}" \ # (you can add as much as you need individually)
    --debug # just to get the output in case something doesn't work
```

If you did everything correctly, you should see a new php binary in `./buildroot/bin`

# EXTRA
If the build works, you should compress your binary with `--with-upx-pack` like this:
```bash
./runtime/php.exe ./bin/spc install-pkg upx # install upx first

./runtime/php.exe ./bin/spc build --with-upx-pack # then your extensions, libs and options
```

Also, what I did before compiling, was listing my current extensions and formatting them into a comma separated list:
```bash
php -m | grep -vE '^\[.*\]|^[A-Z]' | paste -sd, -
```
